### PR TITLE
Generate cloudstack CP CAPI objects

### PR DESCRIPTION
**Pull request by tatlat**
*Issue #, if available:*
[Generate cp capi objects#971](https://github.com/aws/eks-anywhere-internal/issues/971)

*Description of changes:*

*Testing (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/providers/cloudstack/template.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+		return nil, errors.Wrap(err, "generating etcd machine spec")
+	}
+
+	templateBuilder := NewCloudStackTemplateBuilder(
+		&clusterSpec.CloudStackDatacenter.Spec,
+		controlPlaneMachineSpec,
+		etcdMachineSpec,
+		workerNodeGroupMachineSpecs,
+		time.Now,
+	)
```
Comment (cxbrowne1207) [inline pkg/providers/cloudstack/template.go:35]: I think we may need to update the template builder and the way it works. It seems as though it was made prohibiting  control plane and etcd updates, and these provided arguments help to override what's in the provided spec

File: pkg/providers/cloudstack/template.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+func getEtcdMachineSpec(clusterSpec *cluster.Spec) (*v1alpha1.CloudStackMachineConfigSpec, error) {
+	var etcdMachineSpec *v1alpha1.CloudStackMachineConfigSpec
+	if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration != nil {
+		if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef != nil && clusterSpec.CloudStackMachineConfigs[clusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name] != nil {
+			etcdMachineSpec = &clusterSpec.CloudStackMachineConfigs[clusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name].Spec
+		}
+	}
+
+	return etcdMachineSpec, nil
```
Comment (cxbrowne1207) [inline pkg/providers/cloudstack/template.go:47]: Can we return early here if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration is nil?

File: pkg/providers/cloudstack/controlplane.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	"github.com/aws/eks-anywhere/pkg/yamlutil"
+)
+
+// BaseControlPlane represents a CAPI CloudStack control plane.
+type BaseControlPlane = clusterapi.ControlPlane[*cloudstackv1.CloudStackCluster, *cloudstackv1.CloudStackMachineTemplate]
+
+// ControlPlane holds the CloudStack specific objects for a CAPI CloudStack control plane.
+type ControlPlane struct {
+	BaseControlPlane
+	Secrets *corev1.Secret
```
Comment (ddjjia) [inline pkg/providers/cloudstack/controlplane.go:25]: cloudstack cp doesn't have secret. https://github.com/aws/eks-anywhere/blob/main/pkg/providers/cloudstack/config/template-cp.yaml

File: pkg/providers/cloudstack/template.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+		&clusterSpec.CloudStackDatacenter.Spec,
+		controlPlaneMachineSpec,
+		etcdMachineSpec,
+		workerNodeGroupMachineSpecs,
+		time.Now,
+	)
+	return templateBuilder, nil
+}
+
+func getEtcdMachineSpec(clusterSpec *cluster.Spec) (*v1alpha1.CloudStackMachineConfigSpec, error) {
```
Comment (ddjjia) [inline pkg/providers/cloudstack/template.go:39]: Could you return corresponding error?

Comment (ddjjia) [inline pkg/providers/cloudstack/template.go:39]: this only return err as nil

File: pkg/providers/cloudstack/template.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration != nil {
+		if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef != nil && clusterSpec.CloudStackMachineConfigs[clusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name] != nil {
+			etcdMachineSpec = &clusterSpec.CloudStackMachineConfigs[clusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name].Spec
+		}
+	}
+
+	return etcdMachineSpec, nil
+}
+
+func getWorkerNodeGroupMachineSpec(clusterSpec *cluster.Spec) (map[string]v1alpha1.CloudStackMachineConfigSpec, error) {
```
Comment (ddjjia) [inline pkg/providers/cloudstack/template.go:50]: same, only return err as nil

File: pkg/providers/cloudstack/template.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+		if wnConfig.MachineGroupRef != nil && clusterSpec.CloudStackMachineConfigs[wnConfig.MachineGroupRef.Name] != nil {
+			workerNodeGroupMachineSpec = &clusterSpec.CloudStackMachineConfigs[wnConfig.MachineGroupRef.Name].Spec
+			workerNodeGroupMachineSpecs[wnConfig.MachineGroupRef.Name] = *workerNodeGroupMachineSpec
+		}
+	}
+
+	return workerNodeGroupMachineSpecs, nil
+}
+
+func getControlPlaneMachineSpec(clusterSpec *cluster.Spec) (*v1alpha1.CloudStackMachineConfigSpec, error) {
```
Comment (ddjjia) [inline pkg/providers/cloudstack/template.go:63]: same above. Could you also use those extracted function in provider plz? So we don't need to maintain duplicate logic code

File: pkg/providers/cloudstack/controlplane.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tatlat) [reply pkg/providers/cloudstack/controlplane.go]: no it does not, it's been removed

Comment (tatlat) [reply pkg/providers/cloudstack/controlplane.go]: removed it

File: pkg/providers/cloudstack/template.go (hunk 4 above)
Comment (tatlat) [reply pkg/providers/cloudstack/template.go]: i changed it so that now it can return an error

File: pkg/providers/cloudstack/template.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/providers"
+)
+
+func generateTemplateBuilder(clusterSpec *cluster.Spec) (providers.TemplateBuilder, error) {
```
Comment (cxbrowne1207) [inline pkg/providers/cloudstack/template.go:13]: Why not return `CloudStackTemplateBuilder` instead here?

Comment (cxbrowne1207) [inline pkg/providers/cloudstack/template.go:13]: Why not return a CloudStackTemplate builder?

**Pull request merged**